Basic tests and travis CI for the base appengine Dockerfile

dazuma opened the pull request:
Created an initial set of testcases for the base Dockerfile. In particular, tests can take the form of sample applications that inherit from the Dockerfile---the first test is a minimal application that uses only defaults. Further tests can be written later as additional applications, or manually using Ruby/minitest.

Provided a config for running the tests using Travis.

Also included several minor updates to the dockerfile itself: Updated node, removed Ruby 2.2.3, changed the default RACK_ENV, and sped up build by eliminating installation of rdocs.

JustinBeckwith commented:
LGTM

raggi commented:
> GAE Dockerfile updates: update node; remove Ruby 2.2.3; change defaul...

Lets split this out into separate PRs and separate commits.

Aside: Some general advice on commit messages: http://chris.beams.io/posts/git-commit/

raggi commented on appengine/Rakefile:
```diff
@@ -0,0 +1,19 @@
+
+task :build do
+  ::Dir.chdir(::File.dirname(__FILE__)) do
+    puts("docker build -t appengine-ruby-base .")
```
rake has `sh` to do a puts and system, and if you don't provide it a block, i believe it calls fail as expected.

dazuma replied:
Verified and changed to `sh` in https://github.com/GoogleCloudPlatform/ruby-docker/pull/24

raggi commented on appengine/Rakefile:
```diff
@@ -0,0 +1,19 @@
+$VERBOSE = true
```
Instead of setting this globally, maybe you can use rakes `verbose` helper in a specific place - what's this for?

dazuma replied:
Leftover from debugging. Removed in https://github.com/GoogleCloudPlatform/ruby-docker/pull/24

raggi commented on appengine/Rakefile:
```diff
@@ -0,0 +1,19 @@
+  end
+end
+
+task :test => [:build] do
```
Lets use Rake::TestTask, as this breaks things if you provide more tasks to do after test.

dazuma replied:
Done in https://github.com/GoogleCloudPlatform/ruby-docker/pull/24

raggi commented on appengine/test/minimal_app/config.ru:
```diff
@@ -0,0 +1 @@
+run Proc.new { |env| ['200', {'Content-Type' => 'text/plain'}, ["ruby app"]] }
```
The status code should be an integer.

dazuma replied:
Fixed in https://github.com/GoogleCloudPlatform/ruby-docker/pull/24

raggi commented on appengine/test/tc_base_image.rb:
```diff
@@ -0,0 +1,18 @@
+
+class TestBaseImage < ::Minitest::Test
+
+  HELPER = TestHelper.new(verbose: ::ENV['VERBOSE'])
```
This is not picking up the global set in the Rakefile.

Why is the helper instance constant?

If all the tests use a helper like this, we could have a test class that sets this up, something like:

```ruby
class Test < Minitest::Test
  def helper
    @helper ||= Helper.new(ENV['VERBOSE'])
  end
end
```

dazuma replied:
Changed the helper methods to a mixin in https://github.com/GoogleCloudPlatform/ruby-docker/pull/24

raggi commented on appengine/test/tc_base_image.rb:
```diff
@@ -0,0 +1,18 @@
+require "minitest/autorun"
+require "#{::File.dirname(__FILE__)}/test_helper.rb"
```
This is a good case for require_relative.

dazuma replied:
Done in https://github.com/GoogleCloudPlatform/ruby-docker/pull/24

raggi commented on appengine/test/tc_base_image.rb:
```diff
@@ -0,0 +1,18 @@
+require "minitest/autorun"
```
If you do this from `Rake::TestTask` require args, then it won't be necessary to add to every file.

dazuma replied:
I'd rather keep it here. I prefer to be able to run test cases independent of rake. (Unchanged in https://github.com/GoogleCloudPlatform/ruby-docker/pull/24)

raggi replied:
You can:

`ruby -rminitest/autorun test/tc_*.rb`

raggi replied:
At minimum you could move this into test_helper.rb

raggi commented on appengine/test/test_helper.rb:
```diff
@@ -0,0 +1,73 @@
+class TestHelper
+
+  def initialize(opts)
+    @verbose = opts[:verbose]
```
Do we need silent test runs, or can we just always output?

dazuma replied:
Reworked in https://github.com/GoogleCloudPlatform/ruby-docker/pull/24. It always outputs, but simplified (similar to the behavior of rake's `sh`)

raggi commented on appengine/test/test_helper.rb:
```diff
@@ -0,0 +1,73 @@
+      end
+      iter += 1
+      if iter >= timeout
+        raise "Expected \"#{expectation}\" but got \"#{actual}\" (exit code #{exit_code})"
```
If you use minitests `flunk` then you won't get a messy traceback. You may wish to include the cmd in the output.

dazuma replied:
Done in https://github.com/GoogleCloudPlatform/ruby-docker/pull/24

dazuma commented:
Closing. I addressed the above comments by @raggi, squashed, and opened separate pull requests. Those are https://github.com/GoogleCloudPlatform/ruby-docker/pull/24 which does only the tests and rakefile, and https://github.com/GoogleCloudPlatform/ruby-docker/pull/25 which includes travis config and dockerfile speedups. Further dockerfile changes (e.g. nodejs update and RACK_ENV change) will come separately.
